File: src/endpoints/auth/post_ready_login.py
from src import output
from src.libs.db import DB, Table
from . import __types__ as types
from .provider import Provider

async def post_ready_login(params: dict = {}, req = None, _db: DB = None):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.PostReadyLogin(**params)

        # get index
        data = db.get_items(
            table_name=Table.PROVIDER.value,
            unlimited=True,
        )

        # set providers data
        if len(data) > 0:
            def transform_item(item: str) -> dict:
                _result = { 'name': item.get('code') }
                if item.get('code') != 'password':
                    _result['auth_url'] = Provider.get_authorize_url(item.get('code'), params.redirect_uri)
                return _result
            providers = [ transform_item(item) for item in data ]
        else:
            providers = Provider.get_providers()
            def transform_item(provider: str) -> dict:
                _result = { 'name': provider }
                if provider != 'password':
                    _result['auth_url'] = Provider.get_authorize_url(provider, params.redirect_uri)
                return _result
            providers = [ transform_item(provider) for provider in providers ]

        # set result
        result = output.success({
            'message': 'Complete get ready login data.',
            'data': {
                'providers': providers,
            },
        }, _req=req)
    except Exception as e:
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/auth/post-ready-login.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import { PROVIDER_CODE, PROVIDER_TYPE, getProvider } from '@/classes/provider'
import { arrayToObject } from '@/libs/objects'
import type { AuthModel } from './__model'

type PostReadyLogin = {
  body: AuthModel['postReadyLogin']
}

export default async function postReadyLogin({ body }: PostReadyLogin)
{
  try
  {
    // get providers
    const providers = db.getIndex({
      table: DB.TABLE.PROVIDER,
    })
    const _providers = arrayToObject(providers.data, 'code')

    // set providers data
    return Object.values(PROVIDER_CODE).map((code) => {
      if (!_providers[code]) return false
      const __provider__ = getProvider(code)
      if (__provider__.type !== PROVIDER_TYPE.OAUTH) return false
      return {
        name: code,
        auth_url: __provider__.getAuthorizeLink(code, body.redirect_uri),
      }
    }).filter(Boolean)
  }
  catch (_e: any)
  {
    throw new ServiceError('Failed get ready login data.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}
